sanitizers/src/main/java/com/code_intelligence/jazzer/sanitizers/ClojureLangHooks.java

```diff
@@ -0,0 +1,66 @@
+// Copyright 2021 Code Intelligence GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.code_intelligence.jazzer.sanitizers;
+
+import com.code_intelligence.jazzer.api.HookType;
+import com.code_intelligence.jazzer.api.Jazzer;
+import com.code_intelligence.jazzer.api.MethodHook;
+import java.lang.invoke.MethodHandle;
+import java.util.Collections;
+import java.util.Set;
+import java.util.WeakHashMap;
+
+@SuppressWarnings("unused")
+final public class ClojureLangHooks {
+  /**
+   * Used to memoize the all clojure function objects that test for substring. These function
+   * objects extend AFunction and do not overwrite equals(), which allows us to use a WeakHashMap.
+   */
+  static ThreadLocal<Set<Object>> stringContainsFuncs =
+      ThreadLocal.withInitial(() -> Collections.newSetFromMap(new WeakHashMap<Object, Boolean>()));
+
+  /**
+   * This hook checks the type of the returned clojure.lang.IFn objects and puts them into
+   * stringContainsFuncs if they match know string-contains functions.
+   */
+  @MethodHook(
+      type = HookType.AFTER, targetClassName = "clojure.lang.Var", targetMethod = "getRawRoot")
+  public static void
+  clojureMarkContains(
+      MethodHandle method, Object thisObject, Object[] arguments, int hookId, Object result) {
+    if ("clojure.string$includes_QMARK_".equals(result.getClass().getCanonicalName())
```

Comment:
I still think this should be a `Set`, both for readability and (likely, but untested) performance. `Stream.of(...).collect(toSet())` assigned to a `static final` is all the setup that would take.

Reply:
Just out of curiosity I created a benchmark for this:

```java
@State(Scope.Benchmark)
@Fork(value = 1, warmups = 1)
public class TestBenchmark {
  private static final Set<String> haystack =
      Stream.of(
              "clojure.string$includes_QMARK_",
              "clojure.string$starts_with_QMARK_",
              "clojure.string$ends_with_QMARK_",
              "clojure.string$index_of_QMARK_",
              "clojure.string$last_index_of_QMARK_")
          .collect(Collectors.toSet());

  @Param({"foobar", "clojure.string$last_index_of_QMARK_", "clojure.string$last_index_of_QMARK__"})
  String needle;

  @Benchmark
  public boolean equalityChecks() {
    return "clojure.string$includes_QMARK_".equals(needle)
        || "clojure.string$starts_with_QMARK_".equals(needle)
        || "clojure.string$ends_with_QMARK_".equals(needle)
        || "clojure.string$index_of_QMARK_".equals(needle)
        || "clojure.string$last_index_of_QMARK_".equals(needle);
  }

  @Benchmark
  public boolean setContains() {
    return haystack.contains(needle);
  }
}
```

```
Benchmark                                                 (needle)   Mode  Cnt          Score          Error  Units
TestBenchmark.equalityChecks                                foobar  thrpt    5  197581255.803 ±  6287389.319  ops/s
TestBenchmark.equalityChecks   clojure.string$last_index_of_QMARK_  thrpt    5  129550623.215 ±  2934774.598  ops/s
TestBenchmark.equalityChecks  clojure.string$last_index_of_QMARK__  thrpt    5  197707791.842 ±  4276520.409  ops/s
TestBenchmark.setContains                                   foobar  thrpt    5  517792671.251 ± 13372898.282  ops/s
TestBenchmark.setContains      clojure.string$last_index_of_QMARK_  thrpt    5  152513807.705 ±  1303357.002  ops/s
TestBenchmark.setContains     clojure.string$last_index_of_QMARK__  thrpt    5  499847231.177 ±  3257284.885  ops/s
```